pkg/controllers/cluster/taint_manager.go

```diff
@@ -0,0 +1,360 @@
+		if apierrors.IsNotFound(err) {
+			tc.evictionStartTimeMap.delete(keyStr)
+			return nil
+		}
+		return fmt.Errorf("failed to get binding %s: %v", fedKey.NamespaceKey(), err)
+	}
+
+	if !binding.DeletionTimestamp.IsZero() || !util.IsTargetClusterExist(binding.Spec.Clusters, cluster) {
+		tc.evictionStartTimeMap.delete(keyStr)
+		return nil
+	}
+
+	needEviction, tolerationTime, err := tc.needEviction(cluster, binding.Annotations)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if binding needs eviction", "binding", klog.KObj(binding))
+		return err
+	}
+
+	// Case 1: Need eviction now.
+	// Case 2: need eviction after toleration time. If time is up, do eviction right now.
+	// Case 3: Tolerate forever, we do nothing.
+	evictNow := false
+	if needEviction || tolerationTime == 0 {
+		evictNow = true
+	} else if tolerationTime > 0 {
+		now := time.Now()
+		if startTime, exist := tc.evictionStartTimeMap.get(keyStr); exist {
+			// If trigger time has not been reached, just return.
+			if startTime.Add(tolerationTime).After(now) {
+				return nil
+			}
+			// If triggered, evict now.
+			evictNow = true
+		} else {
+			// If first, we set the start time and add this object after tolerationTime.
+			tc.evictionStartTimeMap.set(keyStr, now)
+			tc.bindingEvictionWorker.AddAfter(fedKey, tolerationTime)
+			return nil
+		}
+	}
+
+	if evictNow {
+		if result := util.RemoveTargetCluster(binding.Spec.Clusters, cluster); len(result) < len(binding.Spec.Clusters) {
+			// update final result to evict the target cluster
+			binding.Spec.Clusters = result
+			if err = tc.Update(context.TODO(), binding); err != nil {
+				klog.ErrorS(err, "Failed to update binding", "binding", klog.KObj(binding))
+				return err
+			}
+			tc.emitClusterEvictionEventForResourceBinding(binding, cluster)
+		}
+	}
+	tc.evictionStartTimeMap.delete(keyStr)
+	return nil
+}
+
+func (tc *NoExecuteTaintManager) syncClusterBindingEviction(key util.QueueKey) error {
+	fedKey, ok := key.(keys.FederatedKey)
+	if !ok {
+		klog.Errorf("Failed to sync cluster binding eviction as invalid key: %v", key)
+		return fmt.Errorf("invalid key")
+	}
+	cluster := fedKey.Cluster
+	keyStr := fedKey.String()
+
+	binding := &workv1alpha2.ClusterResourceBinding{}
+	if err := tc.Client.Get(context.TODO(), types.NamespacedName{Name: fedKey.Name}, binding); err != nil {
+		// The resource no longer exist, in which case we stop processing.
+		if apierrors.IsNotFound(err) {
+			tc.evictionStartTimeMap.delete(keyStr)
+			return nil
+		}
+		return fmt.Errorf("failed to get cluster binding %s: %v", fedKey.Name, err)
+	}
+
+	if !binding.DeletionTimestamp.IsZero() || !util.IsTargetClusterExist(binding.Spec.Clusters, cluster) {
+		tc.evictionStartTimeMap.delete(keyStr)
+		return nil
+	}
+
+	needEviction, tolerationTime, err := tc.needEviction(cluster, binding.Annotations)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if cluster binding needs eviction", "binding", binding.Name)
+		return err
+	}
```

Comment:
I guess we can add a short path in case of `needEviction==false`, like:

```golang
	if !needEviction {
		tc.evictionStartTimeMap.delete(keyStr)
		return nil
	}
```

Reply:
I'm afraid not. If `needEviction==false`, maybe the toleration time is not zero, which means  it should be evicted after some period.

Reply:
right.